EIPS/eip-4972.md

```diff
@@ -0,0 +1,143 @@
+    ///  not the current owner. If this function is called again it overwrites the current
+    ///  allowance with _value.
+    function approveFromName(bytes memory _owner, address _spender, uint256 _value) public returns(bool);
+}
+```
+
+### Authentication
+The transfer and approve function is authenticated if and only if the message sender is the owner of the name.
+
+## Rationale
+We use bytes array to represent a name to ensure it’s flexible enough to deal with different use cases. E.g. one can encode the name service contract address the name belongs to into the bytes array. One can also encode extra authentication data, such as zero knowledge proofs, into the bytes array. In the future, we may propose a standard to formalize the name for wider adoption.
+
+The isNameOwner function is sufficient for authenticating the message sender. One can verify the owner by looking up the name owner from a name service, or check zero knowledge proofs encoded in name to prove the ownership directly without looking up anything.
+
+The addressOfName interface decouples the implementation from specific hashing algorithms, as long as the generated address doesn’t collide with EOA/CA address space.
+
+## Backwards Compatibility
+The new account type is compatible with existing ERC token standards.
+
+## Reference Implementation
+### Name Format
+The decoded format of bytes name is not defined at this standard. One straightforward implementation would be:
+```
+bytes memory name = abi.encode((string, ‘address’), (username, nameService))
+```
+where the username is the string representation of the username and nameService is the name service contract address. This will decouple the implementation from specific name services such as ENS.
+
+### Name Derived Address (INameOwnedAccount.addressOfName())
+With the bytes format mentioned above, we can follow the similar rule of CREATE2 opcode to compute the NOA address from nameService and hash of the username as `address(keccak256(0xff, keccak256(“eip-4972.addressOfName”), nameService, keccak256(username)))`.  This can ensure it won’t collide with existing smart contract account addresses.
+
+
+### Ownership of a Name (INameOwnedAccount.isNameOwner())
+Normally we can get the owner from the name service and compare it with the message sender. We recommend the name service to define an owner function in the same format as [ENS.sol](https://github.com/ensdomains/ens/blob/master/contracts/ENS.sol).
+
+### Example Contracts for ERC721 and ERC20 (ERC721NOA and ERC20NOA)
+The example contracts can be found at [NameOwnedAccount.sol](https://github.com/NameOwnedAccount/NameOwnedAccount/blob/main/contracts/identity/NameOwnedAccount.sol), [ERC721NOA.sol](https://github.com/NameOwnedAccount/NameOwnedAccount/blob/main/contracts/token/ERC721/ERC721NOA.sol)[ERC20NOA.sol](https://github.com/NameOwnedAccount/NameOwnedAccount/blob/main/contracts/token/ERC20/ERC20NOA.sol).
```

Comment:
Don't link externally.  Either remove this text if it isn't critical, or if the code is CC0 licensed include it inline or in `../assets/eip-4972/*`

Reply:
Thanks.  Just removed.